src/Http/Http/src/Timeouts/RequestTimeoutsMiddleware.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Http.Timeouts;
+
+internal sealed class RequestTimeoutsMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly IOptionsMonitor<RequestTimeoutOptions> _options;
+
+    public RequestTimeoutsMiddleware(RequestDelegate next, IOptionsMonitor<RequestTimeoutOptions> options)
+    {
+        _next = next;
+        _options = options;
+    }
+
+    public Task Invoke(HttpContext context)
+    {
+        var endpoint = context.GetEndpoint();
+        var timeoutMetadata = endpoint?.Metadata.GetRequiredMetadata<RequestTimeoutAttribute>();
+        var policyMetadata = endpoint?.Metadata.GetRequiredMetadata<RequestTimeoutPolicy>();
+
+        var options = _options.CurrentValue;
+
+        if (timeoutMetadata is null && policyMetadata?.Timeout is null && options.DefaultTimeout is null)
+        {
+            return _next(context);
+        }
+
+        var disableMetadata = endpoint?.Metadata.GetRequiredMetadata<DisableRequestTimeoutAttribute>();
+        if (disableMetadata is not null)
+        {
+            return _next(context);
+        }
+
+        TimeSpan? timespan;
+        if (policyMetadata is not null)
+        {
+            timespan = policyMetadata.Timeout;
+        }
+        else if (timeoutMetadata is not null)
+        {
+            if (timeoutMetadata.Timeout is not null)
+            {
+                timespan = timeoutMetadata.Timeout.Value;
+            }
+            else
+            {
+                if (options.Policies.TryGetValue(timeoutMetadata.PolicyName!, out var policy))
+                {
+                    timespan = policy.Timeout;
+                }
+                else
+                {
+                    // policy does not exists. what to do now!?
+                    throw new Exception();
+                }
+            }
+        }
+        else
+        {
+            timespan = options.DefaultTimeout;
+        }
+
+        if (timespan is null || timespan == Timeout.InfiniteTimeSpan)
+        {
+            return _next(context);
+        }
+
+        return SetTimeoutAsync();
+
+        async Task SetTimeoutAsync()
+        {
+            var originalToken = context.RequestAborted;
+            var timeoutCts = new CancellationTokenSource(timespan.Value);
+            var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(context.RequestAborted, timeoutCts.Token);
+            try
+            {
+                context.RequestAborted = linkedCts.Token;
+                await _next(context);
+            }
+            finally
+            {
```

Comment:
We should consider setting status code to `524` for timed out requests. Otherwise this gets reported as `500`. And it's arguably not an internal server error. Services tend to collect metrics for 5xx codes and 500 are of particular interest. Setting aside issues such as this, where we know nothing wrong had happened, is welcomed.

Reply:
Why do we need to managed another token? Can we just call abort instead.

Reply:
@tekian, the only definition of 524 I can find comes from Cloudflare. 504 seems to be the only official status code related to timeouts. It's not a perfect match since it refers to gateways.

Regardless this will need to be configurable as we discussed elsewhere. I'll work on an amendment to the API for that.

Reply:
Thanks @Tratcher. Configurable & default to `504` is good.